File: 77. Combinations/77. Combinations.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function (n, k) {
  const ans = [];
  const curr = [];

  const backtrack = (start) => {
    if (curr.length == k) {
      ans.push([...curr]);
      return;
    }

    for (let i = start; i < n + 1; i++) {
      curr.push(i);
      backtrack(i + 1);
      curr.pop();
    }
  };

  backtrack(1);
  return ans;
};

File: 77. Combinations/77. Combinations.py
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        ans, curr = [], []

        def backtrack(start):
            if len(curr) == k:
                ans.append(curr[:])
                return

            for i in range(start, n + 1):
                curr.append(i)
                backtrack(i + 1)
                curr.pop()

        backtrack(1)
        return ans
